internal/controller/image/nodeimage_controller.go

```diff
@@ -47,9 +57,44 @@ type NodeImageReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.0/pkg/reconcile
 func (r *NodeImageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := r.Log.WithValues("nodeImage", req.NamespacedName)
+
+	// Fetch the NodeImage instance
+	nodeImage := &imagev1alpha1.NodeImage{}
+	err := r.Get(ctx, req.NamespacedName, nodeImage)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
 
-	// TODO(user): your logic here
+	// Handle deletion
+	if IsDeleted(nodeImage) {
+		log.Info(fmt.Sprintf("NodeImage %s is marked for deletion", nodeImage.Name))
+		// TODO handle deletion
+
+		// Remove finalizer
+		if controllerutil.ContainsFinalizer(nodeImage, NodeImageFinalizer) {
+			controllerutil.RemoveFinalizer(nodeImage, NodeImageFinalizer)
+			if err := r.Update(ctx, nodeImage); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.Info(fmt.Sprintf("Finalizer %s removed from NodeImage %s", NodeImageFinalizer, nodeImage.Name))
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Add finalizer
+	if !controllerutil.ContainsFinalizer(nodeImage, NodeImageFinalizer) {
+		controllerutil.AddFinalizer(nodeImage, NodeImageFinalizer)
+		if err := r.Update(ctx, nodeImage); err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info(fmt.Sprintf("Finalizer %s added to NodeImage %s", NodeImageFinalizer, nodeImage.Name))
+	}
+
+	// TODO handle create/update
```

Comment:
Before utilizing the s3 client in some way here I want to implement the vsphere one first
Apparently it supports pulling from a URL so in that case we can skip storing the image locally
(However, we do need the S3 client later for some cases where URL method is not supported)